### PR TITLE
Include primary groups in user and login filter when restricting group access and also fix user counting in primary groups

### LINKED ISSUE #12190: AD Primary Group issue
I have discovered a problem with the way Primary groups are recognized.  In my example I am going to use the group Domain Users.  In this case, my AD environment all users are a member of Domain Users and this is their Primary group (for 47 of the 48 users).  When I point the OC User Filter to find just these users (Domain Users), it only find 1 user, this happens to be the user who is a member of domain users, but domain users is not this user Primary group.

To confirm, when I go in and start changing users Primary group to something other than Domain Users, the search filter in OC starts to find these users.

Please let me know if we need any logs etc,  In this case the test system is running oc with php 5.5 on Ubuntu 12.02 LTS ownCloud 7.0.3 EE GA

Thanks,

Bob

### PR DESCRIPTION
**Pull request by blizzz**
This is the continuation of https://github.com/owncloud/core/pull/12233 plus user counting fix.

How to test the fixed filter creation? See in the original PR https://github.com/owncloud/core/pull/12233

How to test the user counting? Now, that's a bit more tricky. Partly, I used this script for this: https://gist.github.com/blizzz/5747a97f9332c392a81c
1. have an LDAP configuration towards and AD
2. run this script like `sudo -u www-data ./primGrGetUsers.php "cn=domain users,cn=users,dc=owncloud1,dc=com" s08`  (replace group DN and config prefix accordingly)
3. check whether the output makes sense. My successful looks like this:

user list:
  Alicia(cn=alicia…
  and…
  many…
  more…
  users…
Number of users: 27 took: 1.307079076767
Number of users (counted): 27 took: 0.51463603973389

user list 's…':
  Sergey…
  and…
  3other…
  users…
Number of users with 's…': 4 took: 0.50409507751465
Number of users with 's…' (counted): 4 took: 0.50409507751465
4. Because this is not enough, go to the user management. The number next to the group you are checking should equal the number of users who have this group as primay plus number of users who are members of this as ordinary group
5. You should check the filter now, too. Unfortunately it looks like the filter result numbers for the groups are not updated (in master). However, they are retrieved correctly, you can check with the developer console, network tab and response bodies being logged.

This fixes #12190 and #13533 

request for testing and reviews :) @craigpg @sbelov1 @bboule @MorrisJobke

### CODE REVIEW

**Comment by blizzz** [comment]
There's also a backport for stable7, please test too: https://github.com/owncloud/core/pull/13742

File: apps/user_ldap/group_ldap.php [hunk 1]
```diff
@@ -249,32 +249,71 @@ public function getUserPrimaryGroupIDs($dn) {
...
+		} catch (\Exception $e) {
+			return array();
+		}
+	}
 
-		return $users;
+	/**
+	 * returns the number of users that have the given group as primary group
+	 *
+	 * @param $groupDN
```
Comment (nickvergessen) [inline apps/user_ldap/group_ldap.php]: missing type

File: apps/user_ldap/group_ldap.php [hunk 2]
```diff
@@ -249,32 +249,71 @@ public function getUserPrimaryGroupIDs($dn) {
...
 
 	/**
-	 * returns a list of users that have the given group as primary group
+	 * returns a filter for a "users in primary group" search or count operation
 	 *
-	 * @param string $groupDN
-	 * @param $limit
-	 * @param int $offset
-	 * @return string[]
+	 * @param $groupDN
```
Comment (nickvergessen) [inline apps/user_ldap/group_ldap.php]: missing type

**Comment by DeepDiver1975** [comment]
@MorrisJobke @craigpg @jnfrmarks  this is a bug fix which has to go out with 7.0.5 - as precondition we need this one on OC8 - can I ask you to test these changes with an OC8?

There is also a second pr open #13742 which requires testing with oc7.

Thanks a lot!

File: apps/user_ldap/group_ldap.php [hunk 3]
```diff
@@ -249,32 +249,71 @@ public function getUserPrimaryGroupIDs($dn) {
...
-			$offset
-		);
+	/**
+	 * returns a list of users that have the given group as primary group
+	 *
+	 * @param string $groupDN
+	 * @param string $search
+	 * @param int $limit
+	 * @param int $offset
+	 * @return \string[]
```
Comment (nickvergessen) [inline apps/user_ldap/group_ldap.php]: no backslash needed?

File: apps/user_ldap/group_ldap.php [hunk 4]
```diff
@@ -249,32 +249,71 @@ public function getUserPrimaryGroupIDs($dn) {
...
+	 * @param string $search
+	 * @param int $limit
+	 * @param int $offset
+	 * @return int
+	 */
+	public function countUsersInPrimaryGroup($groupDN, $search = '', $limit = -1, $offset = 0) {
+		try {
+			$filter = $this->prepareFilterForUsersInPrimaryGroup($groupDN, $search);
+			$users = $this->access->countUsers($filter, array('dn'), $limit, $offset);
+			return (is_int($users)) ? $users : 0;
```
Comment (nickvergessen) [inline apps/user_ldap/group_ldap.php]: just `return (int) $users;` ?

Comment (blizzz) [inline apps/user_ldap/group_ldap.php]: or ```intval()``` or  ```+ 0``` … but any of them is better readable than this, true (except + 0 being ugly at the same time).

**Comment by blizzz** [comment]
thx @nickvergessen

**Comment by blizzz** [comment]
@owncloud-bot retest this please

**Comment by DeepDiver1975** [comment]
> @MorrisJobke @craigpg @jnfrmarks this is a bug fix which has to go out with 7.0.5 - as precondition we need this one on OC8 - can I ask you to test these changes with an OC8?
> 
> There is also a second pr open #13742 which requires testing with oc7.
> 
> Thanks a lot!

@MorrisJobke @craigpg @jnfrmarks we are kind of critical with this from a time line perspective - as far as I can tell. We really shall not merge this without proper testing. Let me know what we can do about this. THX

**Comment by jnfrmarks** [comment]
@DeepDiver1975 

I can work on testing this today but may need help on how the LDAP needs to be set up.  The more info that can be put in the issues on how to reproduce the issues that are fixed and the specific ldap setup the easier it will be for me to test.  Is this required to put 8.0 out the door today?

**Comment by MorrisJobke** [comment]
@DeepDiver1975 As it is nothing for 8.0 (issue to fix with this is #12190) I'm fine with moving this to 8.1 but I don't know if @craigpg or @karlitschek wants this in 8.0?

**Comment by MorrisJobke** [comment]
Or we plan it for 8.1 and then maybe backport it to 8.0.1 or something like that.

**Comment by jnfrmarks** [comment]
@MorrisJobke @DeepDiver1975 

It would be helpful for me to know if this needs testing today for 8.0 so I can plan accordingly ;)

**Comment by MorrisJobke** [comment]
@jnfrmarks Sorry. I can't decide on this.

**Comment by blizzz** [comment]
I am personally not so picky whether it goes straight into 8.0 or not, but since this fixes a purple-colored issue (https://github.com/owncloud/core/pull/13742) in OC 7 it would need to get backported anyway.

**Comment by DeepDiver1975** [comment]
> this is a bug fix which has to go out with 7.0.5 - as precondition we need this one on OC8 - can I ask you to test these changes with an OC8?

@MorrisJobke as already explained - we have to fix this within 7.0.5. To follow the workflow we first need to merged into master/OC8.

If we decide to move this to 8.0.1 it will be shipped with 7.0.6 not 7.0.5.



> but since this fixes a purple-colored issue (#13742) in OC 7 it would need to get backported anyway.

there is already a PR open - as soon as you test the change in master you can switch over to the other PR to test it for 7.0.5 - THX

I hope this brings some light into this

**Comment by DeepDiver1975** [comment]
> I can work on testing this today but may need help on how the LDAP needs to be set up. The more info that can be put in the issues on how to reproduce the issues that are fixed and the specific ldap setup the easier it will be for me to test.

@blizzz can you please get in touch with @jnfrmarks and @MorrisJobke to get the required setup sorted out? THX

**Comment by karlitschek** [comment]
@jnfrmarks @craigpg Can you help with testing this please?

**Comment by MorrisJobke** [comment]
> @blizzz can you please get in touch with @jnfrmarks and @MorrisJobke to get the required setup sorted out? THX

I just asked @blizzz : The steps in https://github.com/owncloud/core/pull/12233#issue-49104366 are the testing steps.

**Comment by MorrisJobke** [comment]
Tested with an AD and a LDAP instance and it works :+1:

**Comment by MorrisJobke** [comment]
From @jnfrmarks https://github.com/owncloud/core/pull/12233#issuecomment-72910810 :

> I validated that with AD, the "primaryGroupID" shows up in the raw query for users and that it does not show up in queries for openLDAP. Is there more that needs to be done?

Did you had a look at the group counting? Is it correct?

**Comment by DeepDiver1975** [comment]
pr against master -> milestone 8.1

**Comment by LukasReschke** [comment]
Works, code looks good :+1:

**Comment by blizzz** [comment]
@karlitschek request to backport to stable8. Reference: https://github.com/owncloud/core/issues/13533#issuecomment-100258068 OK?

**Comment by MorrisJobke** [comment]
Backport is in #16456

**Pull request merged**